promql/engine.go

```diff
@@ -606,88 +579,73 @@ func (ev *evaluator) recover(errp *error) {
 	if e == nil {
 		return
 	}
-	if _, ok := e.(runtime.Error); ok {
+	if err, ok := e.(runtime.Error); ok {
 		// Print the stack trace but do not inhibit the running application.
 		buf := make([]byte, 64<<10)
 		buf = buf[:runtime.Stack(buf, false)]
 
 		level.Error(ev.logger).Log("msg", "runtime panic in parser", "err", e, "stacktrace", string(buf))
-		*errp = fmt.Errorf("unexpected error")
+		*errp = fmt.Errorf("unexpected error: %s", err)
 	} else {
 		*errp = e.(error)
 	}
 }
 
-// evalScalar attempts to evaluate e to a Scalar value and errors otherwise.
-func (ev *evaluator) evalScalar(e Expr) Scalar {
-	val := ev.eval(e)
-	sv, ok := val.(Scalar)
-	if !ok {
-		ev.errorf("expected Scalar but got %s", documentedType(val.Type()))
-	}
-	return sv
+func (ev *evaluator) Eval(expr Expr) (v Value, err error) {
+	defer ev.recover(&err)
+	return ev.eval(expr), nil
 }
 
-// evalVector attempts to evaluate e to a Vector value and errors otherwise.
-func (ev *evaluator) evalVector(e Expr) Vector {
-	val := ev.eval(e)
-	vec, ok := val.(Vector)
-	if !ok {
-		ev.errorf("expected instant Vector but got %s", documentedType(val.Type()))
-	}
-	return vec
-}
+// Extra information and caches for evaluating a single node across steps.
+type evalNodeHelper struct {
```

Comment:
Generally seems a bit odd that we have a struct that we pass around all the time. Would it make sense to just directly embed the functionality into the evaluator? Ultimately they are always used together, no?

Reply:
The helper is tied to a node rather than the evaluator, you'd have to wipe it at each node. I've some future optimisations in mind for recording rules where it would end up as part of the node to be re-used across queries.